### ISSUE TITLE
Respect @SuppressWarnings to disable rules

### ISSUE THREAD

**Issue by olafurpg** [scalacenter/scalafix#442]
We recently added support to disable rules over regions of code using comments in #415. Comments were chosen for several reasons,

- they're used the same way in many other linters, including scalastyle, eslint and pylint
- comments allow you to target code more precisely than annotations, since some things cannot be annotated (like a class name, you have to disable the rule for the entire class).
- comments can target code that cannot be annotated, like a group of 10 methods out of 20 methods in a class, here a real world example from the scalafix codebase to disable scalafmt https://github.com/scalacenter/scalafix/blob/d4ccaec624005ce51bc96ee76ea66a871bb4578c/scalafix-core/shared/src/main/scala/scalafix/internal/config/ScalafixReporter.scala#L44-L50)
- comments can be robustly used by syntactic linters (which includes scalafix, it's both syntactic and semantic), while annotations require compilation to be 100% confident the annotation resolved to your expected annotation. A syntactic linter is free to assume that `@SuppressWarnings` means `@java.lang.SuppressWarnings`, but it requires compilation to be 100% guarantee it's true.

However, java.lang.SupppressWarnings is a standard way to suppress messages from the compiler and it's used by linters like WartRemover. I think scalafix should respect `@SuppressWarnings` for those users who still prefer to use annotations.

**Comment by marcelocenerine** [comment]
@olafurpg, what behavior do you expect when comments and annotations are mixed and they overlap?

I took a look at what Checkstyle does, given that it has filters supporting both approaches which can be combined. From what I could see, suppression via annotation takes precedence over comments. Overlapping on/off comments within annotated blocks are still honored but only take effect outside of the enclosing scope. Examples:

- Rule is **disabled** for all lines in `foo1` (annotation) and **enabled** for all lines in `foo2` (comment within `foo1`):
```java
//CHECKSTYLE:OFF
@SuppressWarnings("checkstyle:needBraces")
public void foo1(int n) {
    if (n % 3 == 0) System.out.println("foo");
    //CHECKSTYLE:ON
    if (n % 5 == 0) System.out.println("bar");
}

public void foo2(int n) {
    if (n % 3 == 0) System.out.println("foo");
    if (n % 5 == 0) System.out.println("bar");
}
```

- Rule is **disabled** for all lines in `foo1` (annotation) as well as for `foo2` (comment within `foo1`):
```java
@SuppressWarnings("checkstyle:needBraces")
public void foo1(int n) {
    if (n % 3 == 0) System.out.println("foo");
    //CHECKSTYLE:OFF
    if (n % 5 == 0) System.out.println("bar");
}

public void foo2(int n) {
    if (n % 3 == 0) System.out.println("foo");
    if (n % 5 == 0) System.out.println("bar");
}
```

Are you happy with this behavior for Scalafix? I think it is intuitive.

Also, Checkstyle makes it optional for rule names to be prefixed with "checkstyle:" and accepts `all` to turn all the rules off. The prefix is helpful to know where the rules are coming from. The standard "all" keyword is for convenience and matches the `scalafix:off` switch.

**Comment by olafurpg** [comment]
> Are you happy with this behavior for Scalafix or have a different suggestion?

I totally agree comments should be flat and annotations should be hierarchical.

I hadn't really considered yet how the two systems would mix. One way to do it is to convert annotations into comments so that
```java
@SuppressWarnings("checkstyle:needBraces")
public void foo1(int n) {
    if (n % 3 == 0) System.out.println("foo");
    //CHECKSTYLE:OFF
    if (n % 5 == 0) System.out.println("bar");
}
```

becomes
```java
//CHECKSTYLE:OFF needsBraces
public void foo1(int n) {
    if (n % 3 == 0) System.out.println("foo");
    //CHECKSTYLE:OFF
    if (n % 5 == 0) System.out.println("bar");
} // CHECKSTYLE:ON needsBraces
```

With that translation we would get warnings in `foo2` however. I feel like this system would be easier to document for users. The rule for comments are treated is quite simple

> a rule `r` is disabled in position `p` if there is a comment disabling `r` at position `p1 < p` and there is no comment enabling `r` in position `p2` where `p1 < p2 < p`.

I'm a bit concerned about the complexity of keeping track of two independent layers of off/on comments and annotations.

What do you think?

**Comment by olafurpg** [comment]
I am btw not a big fan of push/pop semantics for comments since I believe it makes it harder to reason why a rule is disabled/enabled at a given location. For example, what happens in cases like here?

```
// off 1
// off 2
// on ?
```
- push/pop: `// on` disables `// off 2` and all rules are still disabled because of `// off 1`. You would need two `// on` to enable scalafix for the rest of the file, which looks weird.
- flat (current scalafix rule): `// on` disables both `// off 1` and `// off 2`.

**Comment by marcelocenerine** [comment]
>One way to do it is to convert annotations into comments so that

While ago I took a look at this issue and converting annotations into on/off toggles was my first tackle at it. I got it partially working but then got some non-intuitive results while testing: one is the scenario you described above where `foo2` would get warnings given that the `OFF` comment within the annotated method gets dropped as it is redundant (scalafix points that out though). The other scenario that I can remember is the following:

```java
 //CHECKSTYLE:OFF
@SuppressWarnings("needBraces")
public void foo1(int n) {
    if (n % 3 == 0) System.out.println("foo");
    if (n % 5 == 0) System.out.println("bar");
}

public void foo2(int n) {
    if (n % 3 == 0) System.out.println("foo");
    if (n % 5 == 0) System.out.println("bar");
}
```

If we converted the annotation into comments with no distinction, scalafix would drop its corresponding `OFF` switch as it is redundant with the comment above. However, the `ON` switch after exiting the method would remain, which would incorrectly (debatable) turn the rules back to ON again. After these experiments I came to the conclusion that I was trying to compare apples and oranges.

In practice I bet people wouldn't write such combinations as these would just cause confusion and should be highly discouraged. However, if scalafix allow such combinations then it would be reasonable to expect them to be handled gracefully.

> I am btw not a big fan of push/pop semantics for comments since I believe it makes it harder to reason why a rule is disabled/enabled at a given location.

I think a push/pop style is intuitive from the hierarchical perspective. Checkstyle seems to combine both styles nicely.

**Comment by olafurpg** [comment]
That is a good example @marcelocenerine nice catch.

> I think a push/pop style is intuitive from the hierarchical perspective. Checkstyle seems to combine both styles nicely.

I agree, it seems push/pop is more logical for annotations because by construction we are guaranteed that every ON has a matching OFF and there are no overlapping ranges.

WDYT @MasseGuillaume ?

Some other question

- should scalafix report "Unused SupressWarning" for redundant `@SuppressWarning("scalafix:UnusedTerm")`? I think it makes sense as long as string literal starts with "scalafix: ", otherwise we should drop it.
- should `@SuppressWarning("UnusedTerm")` disable the UnusedTerm rule?
  - IMO it should be disabled.
- should `@SuppressWarning("all")` disable all rules like `// scalafix: off`?
  - IMO I think it should disable all rules.

**Comment by marcelocenerine** [comment]
> push/pop is more logical for annotations because by construction we are guaranteed that every ON has a matching OFF and there are no overlapping ranges.

That's a perfect way to describe it :)

**Comment by MasseGuillaume** [comment]
I think

```
// off 1
// off 2
// on ?
```

on should put everything back on.

annotation's positions are less precise than comments, I would put precedence on comments

**Comment by marcelocenerine** [comment]
>annotation's positions are less precise than comments, I would put precedence on comments

@MasseGuillaume If precedence is given to comments, then things like in the following example won't work as users might expect (me included):

```scala
  // scalafix:off EscapeHatchDummyLinter
  def aDummy = {}
  // scalafix:on EscapeHatchDummyLinter

  @SuppressWarnings(Array("EscapeHatchDummyLinter"))
  def bDummy = {}

  @SuppressWarnings(Array("EscapeHatchDummyLinter"))
  def cDummy = {}
```

because `scalafix:on` turns the rule ON after `aDummy`, the rule stays active for `bDummy` and `cDummy`, even though the latter methods have the `@SuppressWarnings` annotation. I think it makes more sense to do what Checkstyle does and give precedence to the annotation.

Btw, I'm nearly done with this feature. I will post a PR at some point this week.

**Comment by olafurpg** [comment]
Fixed in #620
